File: packages/alpaca-option/src/payoff.ts
import { fail } from './error';
import type { PayoffLegInput } from './types';

const ROOT_EPSILON = 1e-9;

function ensureFinite(code: string, name: string, value: number): void {
  if (!Number.isFinite(value)) {
    fail(code, `${name} must be finite: ${value}`);
  }
}

function ensurePositive(code: string, name: string, value: number): void {
  ensureFinite(code, name, value);
  if (value <= 0) {
    fail(code, `${name} must be greater than zero: ${value}`);
  }
}

function validateLeg(leg: PayoffLegInput): void {
  ensurePositive('invalid_payoff_input', 'strike', leg.strike);
  ensureFinite('invalid_payoff_input', 'premium', leg.premium);
  if (!Number.isInteger(leg.quantity) || leg.quantity <= 0) {
    fail('invalid_payoff_input', `quantity must be a positive integer: ${leg.quantity}`);
  }
}

function legIntrinsic(leg: PayoffLegInput, underlyingPriceAtExpiry: number): number {
  return leg.optionRight === 'call'
    ? Math.max(underlyingPriceAtExpiry - leg.strike, 0)
    : Math.max(leg.strike - underlyingPriceAtExpiry, 0);
}

function signedLegPayoff(leg: PayoffLegInput, underlyingPriceAtExpiry: number): number {
  const intrinsic = legIntrinsic(leg, underlyingPriceAtExpiry);
  return leg.positionSide === 'long'
    ? leg.quantity * (intrinsic - leg.premium)
    : leg.quantity * (leg.premium - intrinsic);
}

function legSlope(leg: PayoffLegInput, underlyingPriceAtExpiry: number): number {
  if (leg.optionRight === 'call') {
    if (underlyingPriceAtExpiry > leg.strike) {
      return leg.positionSide === 'long' ? leg.quantity : -leg.quantity;
    }
    return 0;
  }

  if (underlyingPriceAtExpiry < leg.strike) {
    return leg.positionSide === 'long' ? -leg.quantity : leg.quantity;
  }
  return 0;
}

function maybePushRoot(
  legs: PayoffLegInput[],
  start: number,
  end: number,
  sample: number,
  roots: number[],
): void {
  const slope = legs.reduce((total, leg) => total + legSlope(leg, sample), 0);
  const value = strategyPayoffAtExpiry({ legs, underlyingPriceAtExpiry: sample });
  if (Math.abs(slope) <= ROOT_EPSILON) {
    return;
  }

  const root = sample - value / slope;
  if (!Number.isFinite(root) || root < 0) {
    return;
  }

  const startOk = root + ROOT_EPSILON >= start;
  const endOk = Number.isFinite(end) ? root - ROOT_EPSILON <= end : true;
  if (startOk && endOk) {
    roots.push(root);
  }
}

export function singleLegPayoffAtExpiry(input: {
  optionRight: PayoffLegInput['optionRight'];
  positionSide: PayoffLegInput['positionSide'];
  strike: number;
  premium: number;
  quantity: number;
  underlyingPriceAtExpiry: number;
}): number {
  return strategyPayoffAtExpiry({
    legs: [{
      optionRight: input.optionRight,
      positionSide: input.positionSide,
      strike: input.strike,
      premium: input.premium,
      quantity: input.quantity,
    }],
    underlyingPriceAtExpiry: input.underlyingPriceAtExpiry,
  });
}

export function strategyPayoffAtExpiry(input: {
  legs: PayoffLegInput[];
  underlyingPriceAtExpiry: number;
}): number {
  ensureFinite('invalid_payoff_input', 'underlyingPriceAtExpiry', input.underlyingPriceAtExpiry);
  if (input.underlyingPriceAtExpiry < 0) {
    fail('invalid_payoff_input', `underlyingPriceAtExpiry must be non-negative: ${input.underlyingPriceAtExpiry}`);
  }

  let total = 0;
  for (const leg of input.legs) {
    validateLeg(leg);
    total += signedLegPayoff(leg, input.underlyingPriceAtExpiry);
  }
  return total;
}

export function breakEvenPoints(input: { legs: PayoffLegInput[] }): number[] {
  for (const leg of input.legs) {
    validateLeg(leg);
  }
  if (input.legs.length === 0) {
    return [];
  }

  const strikes = [...new Set(input.legs.map((leg) => leg.strike).sort((a, b) => a - b))];
  const roots: number[] = [];
  let intervalStart = 0;

  strikes.forEach((boundary, index) => {
    const sample = index === 0 ? Math.max(boundary / 2, 0) : (intervalStart + boundary) / 2;
    maybePushRoot(input.legs, intervalStart, boundary, sample, roots);
    intervalStart = boundary;
  });

  maybePushRoot(input.legs, intervalStart, Number.POSITIVE_INFINITY, intervalStart + Math.max(intervalStart, 1), roots);
  return [...new Set(roots.map((value) => Math.round(value * 1e8) / 1e8))].sort((a, b) => a - b);
}

File: crates/alpaca-option/src/payoff.rs
use crate::error::{OptionError, OptionResult};
use crate::types::{PayoffLegInput, PositionSide};

const ROOT_EPSILON: f64 = 1e-9;

fn ensure_finite(code: &'static str, name: &str, value: f64) -> OptionResult<()> {
    if value.is_finite() {
        Ok(())
    } else {
        Err(OptionError::new(
            code,
            format!("{name} must be finite: {value}"),
        ))
    }
}

fn ensure_positive(code: &'static str, name: &str, value: f64) -> OptionResult<()> {
    ensure_finite(code, name, value)?;
    if value > 0.0 {
        Ok(())
    } else {
        Err(OptionError::new(
            code,
            format!("{name} must be greater than zero: {value}"),
        ))
    }
}

fn validate_leg(leg: &PayoffLegInput) -> OptionResult<()> {
    ensure_positive("invalid_payoff_input", "strike", leg.strike)?;
    ensure_finite("invalid_payoff_input", "premium", leg.premium)?;
    Ok(())
}

fn leg_intrinsic(leg: &PayoffLegInput, underlying_price_at_expiry: f64) -> f64 {
    match leg.option_right.as_str() {
        "call" => (underlying_price_at_expiry - leg.strike).max(0.0),
        "put" => (leg.strike - underlying_price_at_expiry).max(0.0),
        _ => 0.0,
    }
}

fn signed_leg_payoff(leg: &PayoffLegInput, underlying_price_at_expiry: f64) -> f64 {
    let intrinsic = leg_intrinsic(leg, underlying_price_at_expiry);
    let quantity = leg.quantity as f64;
    match leg.position_side {
        PositionSide::Long => quantity * (intrinsic - leg.premium),
        PositionSide::Short => quantity * (leg.premium - intrinsic),
    }
}

fn leg_slope(leg: &PayoffLegInput, underlying_price_at_expiry: f64) -> f64 {
    let quantity = leg.quantity as f64;
    match (leg.option_right.as_str(), &leg.position_side) {
        ("call", PositionSide::Long) => {
            if underlying_price_at_expiry > leg.strike {
                quantity
            } else {
                0.0
            }
        }
        ("call", PositionSide::Short) => {
            if underlying_price_at_expiry > leg.strike {
                -quantity
            } else {
                0.0
            }
        }
        ("put", PositionSide::Long) => {
            if underlying_price_at_expiry < leg.strike {
                -quantity
            } else {
                0.0
            }
        }
        ("put", PositionSide::Short) => {
            if underlying_price_at_expiry < leg.strike {
                quantity
            } else {
                0.0
            }
        }
        _ => 0.0,
    }
}

fn maybe_push_root(
    legs: &[PayoffLegInput],
    start: f64,
    end: f64,
    sample: f64,
    roots: &mut Vec<f64>,
) -> OptionResult<()> {
    let slope = legs.iter().map(|leg| leg_slope(leg, sample)).sum::<f64>();
    let value = strategy_payoff_at_expiry(legs, sample)?;
    if slope.abs() <= ROOT_EPSILON {
        return Ok(());
    }

    let root = sample - value / slope;
    if !root.is_finite() || root < 0.0 {
        return Ok(());
    }

    let start_ok = root + ROOT_EPSILON >= start;
    let end_ok = if end.is_finite() {
        root - ROOT_EPSILON <= end
    } else {
        true
    };
    if start_ok && end_ok {
        roots.push(root);
    }
    Ok(())
}

pub fn single_leg_payoff_at_expiry(
    option_right: &str,
    position_side: &str,
    strike: f64,
    premium: f64,
    quantity: u32,
    underlying_price_at_expiry: f64,
) -> OptionResult<f64> {
    let leg = PayoffLegInput::new(option_right, position_side, strike, premium, quantity)?;
    strategy_payoff_at_expiry(&[leg], underlying_price_at_expiry)
}

pub fn strategy_payoff_at_expiry(
    legs: &[PayoffLegInput],
    underlying_price_at_expiry: f64,
) -> OptionResult<f64> {
    ensure_finite(
        "invalid_payoff_input",
        "underlying_price_at_expiry",
        underlying_price_at_expiry,
    )?;
    if underlying_price_at_expiry < 0.0 {
        return Err(OptionError::new(
            "invalid_payoff_input",
            format!(
                "underlying_price_at_expiry must be non-negative: {underlying_price_at_expiry}"
            ),
        ));
    }

    let mut total = 0.0;
    for leg in legs {
        validate_leg(leg)?;
        total += signed_leg_payoff(leg, underlying_price_at_expiry);
    }
    Ok(total)
}

pub fn break_even_points(legs: &[PayoffLegInput]) -> OptionResult<Vec<f64>> {
    for leg in legs {
        validate_leg(leg)?;
    }
    if legs.is_empty() {
        return Ok(Vec::new());
    }

    let mut strikes = legs.iter().map(|leg| leg.strike).collect::<Vec<_>>();
    strikes.sort_by(|left, right| left.partial_cmp(right).unwrap());
    strikes.dedup_by(|left, right| (*left - *right).abs() <= ROOT_EPSILON);

    let mut roots = Vec::new();
    let mut interval_start = 0.0;
    for (index, boundary) in strikes.iter().enumerate() {
        let sample = if index == 0 {
            (boundary / 2.0).max(0.0)
        } else {
            (interval_start + boundary) / 2.0
        };
        maybe_push_root(legs, interval_start, *boundary, sample, &mut roots)?;
        interval_start = *boundary;
    }

    let tail_sample = interval_start + interval_start.max(1.0);
    maybe_push_root(legs, interval_start, f64::INFINITY, tail_sample, &mut roots)?;

    roots.sort_by(|left, right| left.partial_cmp(right).unwrap());
    roots.dedup_by(|left, right| (*left - *right).abs() <= 1e-7);
    Ok(roots)
}
